src/fswAlgorithms/attGuidance/oneAxisSolarArrayPoint/oneAxisSolarArrayPoint.c

```diff
@@ -160,4 +160,239 @@ void Update_oneAxisSolarArrayPoint(oneAxisSolarArrayPointConfig *configData, uin
+{
+    double PRV_theta[3];
+
+    if (alignmentPriority == prioritizeAxisAlignment) {
+        for (int i = 0; i < 3; i++) {
+            PRV_theta[i] = 0;
+        }
+    }
+    else {
+        double sTheta = v3Dot(rHat_SB_R2, a1Hat_B);
+        double theta = asin( fmin( fmax( fabs(sTheta), -1 ), 1 ) );
+        if (fabs(theta) < EPS) {
+            // if Sun direction and solar array drive are already perpendicular, third rotation is null
+            for (int i = 0; i < 3; i++) {
+            PRV_theta[i] = 0;
+            }
+        }
+        else {
+            // if Sun direction and solar array drive are not perpendicular, project solar array drive a1Hat_B onto perpendicular plane (aPHat_B) and compute third rotation
+            double e_theta[3], aPHat_B[3];
+            if (fabs(fabs(theta)-MPI/2) > EPS) {
+                for (int i = 0; i < 3; i++) {
+                    aPHat_B[i] = (a1Hat_B[i] - sTheta * rHat_SB_R2[i]) / (1 - sTheta * sTheta);
+                }
+                v3Cross(a1Hat_B, aPHat_B, e_theta);
+            }
+            else {
+                // rotate about the axis that minimizes variation in hRefHat_B direction
+                v3Cross(rHat_SB_R2, hRefHat_B, aPHat_B);
+                if (v3Norm(aPHat_B) < EPS) {
+                    v3Perpendicular(rHat_SB_R2, aPHat_B);
+                }
+                v3Cross(a1Hat_B, aPHat_B, e_theta);
+            }
+            v3Normalize(e_theta, e_theta);
+            v3Scale(theta, e_theta, PRV_theta);
+        }
+    }
+
+    /*! compute third rotation R3R2 */
+    PRV2C(PRV_theta, R3R2);
+}
+
+/*! This helper function computes the final rotation as a product of the first three DCMs */
+void computeFinalRotation(int alignmentPriority, double BN[3][3], double rHat_SB_B[3], double hRefHat_B[3], double hReqHat_B[3], double a1Hat_B[3], double a2Hat_B[3], double RN[3][3])
+{
+    /*! compute the first rotation DCM */
+    double R1B[3][3];
+    computeFirstRotation(hRefHat_B, hReqHat_B, R1B);
+
+    /*! compute Sun direction vector in R1 frame coordinates */
+    double rHat_SB_R1[3];
+    m33MultV3(R1B, rHat_SB_B, rHat_SB_R1);
+
+    /*! compute the second rotation DCM */
+    double R2R1[3][3];
+    computeSecondRotation(hRefHat_B, rHat_SB_R1, a1Hat_B, a2Hat_B, R2R1);
+
+    /* compute Sun direction in R2 frame components */
+    double rHat_SB_R2[3];
+    m33MultV3(R2R1, rHat_SB_R1, rHat_SB_R2);
+
+    /*! compute the third rotation DCM */
+    double R3R2[3][3];
+    computeThirdRotation(alignmentPriority, hRefHat_B, rHat_SB_R2, a1Hat_B, R3R2);
+
+    /*! compute reference frames w.r.t inertial frame */
+    double R1N[3][3], R2N[3][3];
+    m33MultM33(R1B, BN, R1N);
+    m33MultM33(R2R1, R1N, R2N);
+    m33MultM33(R3R2, R2N, RN);
+}
+
+/*! This helper function outputs a generic vector perpendicular to the input vector */
+void v3Perpendicular(double x[3], double y[3])
```

Comment:
Should we add something like this to linearAlgebra in the future?

Reply:
Added to the linearAlgebra library